internal/controller/services/gateway/gateway_controller_actions.go

```diff
@@ -0,0 +1,180 @@
+		"gateway", "odh-gateway",
+		"namespace", gatewayInstance.Spec.Namespace,
+		"domain", gatewayInstance.Spec.Domain,
+		"certificateType", gatewayInstance.Spec.Certificates.Type)
+
+	return nil
+}
+
+func createGatewayClass(rr *odhtypes.ReconciliationRequest) error {
+	gatewayClass := &gwapiv1.GatewayClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "odh-gateway-class",
+			Labels: map[string]string{
+				"app.kubernetes.io/managed-by": "opendatahub-operator",
+				"opendatahub.io/internal":      "true",
+			},
+		},
+		Spec: gwapiv1.GatewayClassSpec{
+			ControllerName: "openshift.io/gateway-controller/v1",
+		},
+	}
+
+	return rr.AddResources(gatewayClass)
+}
+
+func handleCertificates(ctx context.Context, rr *odhtypes.ReconciliationRequest, gatewayInstance *serviceApi.Gateway) (string, error) {
+	l := logf.FromContext(ctx).WithName("handleCertificates")
+
+	certType := gatewayInstance.Spec.Certificates.Type
+
+	switch certType {
+	case "cert-manager":
+		return createCertManagerCertificate(rr, gatewayInstance)
+	case "user-provided":
+		l.Info("User-provided certificates not yet implemented")
+		return "", nil
+	case "openshift-service-ca":
+		l.Info("OpenShift service CA not yet implemented")
+		return "", nil
+	default:
+		l.Info("No certificate type specified, HTTP only", "type", certType)
+		return "", nil
+	}
+}
+
+func createCertManagerCertificate(rr *odhtypes.ReconciliationRequest, gatewayInstance *serviceApi.Gateway) (string, error) {
+	secretName := fmt.Sprintf("%s-tls", gatewayInstance.Name)
+
+	issuerRef := cmmeta.ObjectReference{
+		Name: "selfsigned-cluster-issuer",
+		Kind: "ClusterIssuer",
+	}
+
+	if gatewayInstance.Spec.Certificates.IssuerRef != nil {
+		issuerRef = *gatewayInstance.Spec.Certificates.IssuerRef
+	}
+
+	certificate := &certmanagerv1.Certificate{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-cert", gatewayInstance.Name),
+			Namespace: gatewayInstance.Spec.Namespace,
+			Labels: map[string]string{
+				"app.kubernetes.io/managed-by": "opendatahub-operator",
+				"opendatahub.io/internal":      "true",
+			},
+		},
+		Spec: certmanagerv1.CertificateSpec{
+			SecretName: secretName,
+			DNSNames:   []string{gatewayInstance.Spec.Domain},
+			IssuerRef:  issuerRef,
+		},
+	}
+
+	if err := rr.AddResources(certificate); err != nil {
+		return "", fmt.Errorf("failed to add Certificate resource: %w", err)
+	}
+
+	return secretName, nil
+}
+
+func createGateway(rr *odhtypes.ReconciliationRequest, gatewayInstance *serviceApi.Gateway, certSecretName string) error {
+	listeners := createListeners(certSecretName)
+
+	gateway := &gwapiv1.Gateway{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "odh-gateway",
+			Namespace: gatewayInstance.Spec.Namespace,
+			Labels: map[string]string{
+				"app.kubernetes.io/managed-by": "opendatahub-operator",
+				"opendatahub.io/internal":      "true",
+			},
+		},
+		Spec: gwapiv1.GatewaySpec{
+			GatewayClassName: "odh-gateway-class",
```

Comment:
@coderabbitai `odh-gateway-class` is used in more than one place; is it better to declare it as a constant?